### PR TITLE
Reenable clippy testing

### LINKED ISSUE #45680: Compile error: clippy_lints does not support feature(macro_vis_matcher)
branch: master 
   Compiling cargo_metadata v0.2.3
   Compiling clippy_lints v0.0.166 (file:///home/ryan/git/rust/src/tools/clippy/clippy_lints)
error: :vis fragment specifier is experimental and subject to change (see issue #41022)
 --> <declare_lint macros>:1:3
  |
1 | ( $ vis : vis $ NAME : ident , $ Level : ident , $ desc : expr ) => (
  |   ^^^^^^^^^^^
  |
  = help: add #![feature(macro_vis_matcher)] to the crate attributes to enable

error: :vis fragment specifier is experimental and subject to change (see issue #41022)
 --> <declare_lint macros>:1:3
  |
1 | ( $ vis : vis $ NAME : ident , $ Level : ident , $ desc : expr ) => (
  |   ^^^^^^^^^^^
  |
  = help: add #![feature(macro_vis_matcher)] to the crate attributes to enable

error: aborting due to previous error

error: aborting due to previous error

error: Could not compile `clippy_lints`.
warning: build failed, waiting for other jobs to finish...
error: Could not compile `clippy_lints`.
warning: build failed, waiting for other jobs to finish...
error: build failed
This failure is expected (see `src/tools/toolstate.toml`)
Build completed successfully in 1:46:49

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #45680

### CODE REVIEW

**Comment by kennytm** [comment]
This should not land before #45247 (breaks rustfmt/rls) is merged.

**Comment by kennytm** [comment]
@oli-obk #45427 is merged, could you verify if clippy can still be compiled or require another update? Thanks!

**Comment by oli-obk** [comment]
It required another update. I rebased over master and amended the commit.

**Comment by oli-obk** [comment]
Ping @kennytm 

Not sure if it still builds though.

**Comment by kennytm** [comment]
Two macOS builds (and only macs, strangely) failed. Looks like you do need to update again. 

```
[01:19:46] error[E0608]: cannot index into a value of type `std::rc::Rc<std::collections::HashSet<rustc::hir::ItemLocalId, std::hash::BuildHasherDefault<rustc_data_structures::fx::FxHasher>>>`
[01:19:46]    --> src/tools/clippy/clippy_lints/src/methods.rs:893:26
[01:19:46]     |
[01:19:46] 893 |         let promotable = cx.tcx.rvalue_promotable_map(owner_def)[&arg.hir_id.local_id];
[01:19:46]     |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[01:19:46] 
[01:19:47] error: aborting due to previous error
[01:19:47] 
[01:19:47] error: Could not compile `clippy_lints`.
[01:19:47] warning: build failed, waiting for other jobs to finish...
[01:19:55] error: build failed
```

File: src/tools/toolstate.toml [hunk 1]
```diff
@@ -26,7 +26,7 @@
 miri = "Broken"
 
 # ping @Manishearth @llogiq @mcarton @oli-obk
-clippy = "Broken"
+clippy = "Compiling"
```
Comment (kennytm) [inline src/tools/toolstate.toml]: Is it intended it is changed to `Compiling` not `Testing`?

Comment (oli-obk) [reply src/tools/toolstate.toml]: Yea, testing wasn't allowed until a few hours ago: https://github.com/nrc/dev-tools-team/issues/18#issuecomment-343050790

I'll rebase to fix the failure and enable testing

**Comment by oli-obk** [comment]
Clippy can now be tested. There are some weird issues with compiletest, but I'm thinking we can probably rig rustc's compiletest to run the tests instead

**Comment by kennytm** [comment]
Sorry clippy needs to be updated again 😞 

```
[01:23:56]    Compiling clippy_lints v0.0.170 (file:///Users/travis/build/rust-lang/rust/src/tools/clippy/clippy_lints)
[01:23:58] error[E0432]: unresolved import `rustc::mir::transform`
[01:23:58]   --> src/tools/clippy/clippy_lints/src/utils/mod.rs:12:17
[01:23:58]    |
[01:23:58] 12 | use rustc::mir::transform::MirSource;
[01:23:58]    |                 ^^^^^^^^^ Could not find `transform` in `mir`
[01:23:58] 
[01:24:02] error[E0432]: unresolved import `rustc::mir::transform`
[01:24:02]   --> src/tools/clippy/clippy_lints/src/utils/mod.rs:12:17
[01:24:02]    |
[01:24:02] 12 | use rustc::mir::transform::MirSource;
[01:24:02]    |                 ^^^^^^^^^ Could not find `transform` in `mir`
[01:24:02] 
[01:24:02] error: aborting due to previous error
[01:24:02] 
[01:24:02] error: Could not compile `clippy_lints`.
[01:24:02] warning: build failed, waiting for other jobs to finish...
[01:24:06] error: aborting due to previous error
[01:24:06] 
[01:24:06] error: Could not compile `clippy_lints`.
```

**Comment by kennytm** [comment]
@bors retry 

Corrupted submodule??

```
[00:03:45] fatal: packed object 5b3f5468043e7ab6883fa7683e795aadf311017f (stored in /Users/travis/rustsrc/src/.git/modules/src/doc/nomicon/objects/pack/pack-8b46ddbde06b2411f3957d8dade78fe90ec97c01.pack) is corrupt
[00:03:45] Submodule path 'src/doc/reference': checked out '36adc6ae504c6e0343ab5d7b3871f0a2a71236da'
[00:03:46] Submodule path 'src/jemalloc': checked out '1f5a28755e301ac581e2048011e4e0ff3da482ef'
[00:03:46] Submodule path 'src/libcompiler_builtins': checked out '0b9844764ea1f99ea11a7917a4f3ba7fd2db775c'
[00:03:46] Submodule path 'src/liblibc': checked out 'c1068cd82ae55907e8fd4457b98278a3aaa9162b'
[00:03:46] Submodule path 'src/rt/hoedown': checked out 'da282f1bb7277b4d30fa1599ee29ad8eb4dd2a92'
[00:03:47] Submodule path 'src/tools/cargo': checked out 'abd137ad14a8f352b48f24b11c3995d716deb260'
[00:03:49] Submodule path 'src/tools/clippy': checked out 'f76225e3887170743403af9204887918b5db5a80'
[00:03:51] Submodule path 'src/tools/miri': checked out '80853e2f24a01db96fe9821e468dd2af75a4d2e5'
[00:03:51] Submodule path 'src/tools/rls': checked out '015073b7c8a192584c15b2c1bd34a92025cf6836'
[00:03:51] Submodule path 'src/tools/rust-installer': checked out '0ddd53c4bc2a76df565a1c1fc0cc6f19f254b51e'
[00:03:54] Submodule path 'src/tools/rustfmt': checked out '291dbe8d87701f3617ccdd70caa49df1c27af0e0'
[00:03:54] Unable to checkout '1625e0b8c870891b84b0969777a974bf87be579b' in submodule path 'src/doc/nomicon'
```

**Pull request merged**